plugin/evm/acp176/acp176.go

```diff
@@ -0,0 +1,197 @@
+	TargetConversion    = MaxTargetChangeRate * MaxTargetExcessDiff // D
+	MaxTargetExcessDiff = 1 << 15                                   // Q
+	MinBaseFee          = 1                                         // M
+
+	TimeToFillCapacity            = 10   // in seconds
+	TargetToMax                   = 2    // multiplier to convert from target per second to max per second
+	TargetToPriceUpdateConversion = 43   // 43s ~= 30s * ln(2) which makes the price double at most every ~30 seconds
+	MaxTargetChangeRate           = 1024 // Controls the rate that the target can change per block.
+
+	targetToMaxCapacity = TargetToMax * TimeToFillCapacity
+	maxTargetExcess     = 1_024_950_627 // TargetConversion * ln(MaxUint64 / MinTargetPerSecond) + 1
+)
+
+var minBaseFee = big.NewInt(MinBaseFee)
+
+// State represents the current state of the gas pricing and constraints.
+type State struct {
+	Gas          gas.State
+	TargetExcess gas.Gas
+}
+
+// Target returns the target gas consumed per second.
+//
+// Target = MinTargetPerSecond * e^(TargetExcess / TargetConversion)
+func (s *State) Target() gas.Gas {
+	return gas.Gas(gas.CalculatePrice(
+		MinTargetPerSecond,
+		s.TargetExcess,
+		TargetConversion,
+	))
+}
+
+// MaxCapacity returns the maximum possible accrued gas capacity.
+func (s *State) MaxCapacity() gas.Gas {
+	targetPerSecond := s.Target()
+	maxCapacity, err := safemath.Mul(targetToMaxCapacity, targetPerSecond)
+	if err != nil {
+		maxCapacity = math.MaxUint64
+	}
+	return maxCapacity
+}
+
+// BaseFee returns the current required fee per gas.
+//
+// BaseFee = MinBaseFee * e^(Excess / (Target() * TargetToPriceUpdateConversion))
+func (s *State) BaseFee() *big.Int {
+	target := s.Target()
+	priceUpdateConversion, err := safemath.Mul(TargetToPriceUpdateConversion, target)
+	if err != nil {
+		priceUpdateConversion = math.MaxUint64
+	}
+
+	bigExcess := new(big.Int).SetUint64(uint64(s.Gas.Excess))
+	bigPriceUpdateConversion := new(big.Int).SetUint64(uint64(priceUpdateConversion))
+	return fakeExponential(minBaseFee, bigExcess, bigPriceUpdateConversion)
+}
+
+// AdvanceTime increases the gas capacity and decreases the gas excess based on
+// the elapsed seconds.
+func (s *State) AdvanceTime(seconds uint64) {
+	targetPerSecond := s.Target()
+	maxPerSecond, err := safemath.Mul(TargetToMax, targetPerSecond)
+	if err != nil {
+		maxPerSecond = math.MaxUint64
+	}
+	maxCapacity, err := safemath.Mul(TimeToFillCapacity, maxPerSecond)
+	if err != nil {
+		maxCapacity = math.MaxUint64
+	}
+	s.Gas = s.Gas.AdvanceTime(
+		maxCapacity,
+		maxPerSecond,
+		targetPerSecond,
+		seconds,
+	)
+}
+
+// ConsumeGas decreases the gas capacity and increases the gas excess by
+// gasUsed + extraGasUsed. If the gas capacity is insufficient, an error is
+// returned.
+func (s *State) ConsumeGas(
+	gasUsed uint64,
+	extraGasUsed *big.Int,
+) error {
+	var err error
+	s.Gas, err = s.Gas.ConsumeGas(gas.Gas(gasUsed))
```

Comment:
My old readability mentor (then the head of el Goog readability team) does a much better job than me at explaining why he thinks [this is an antipattern](https://gist.github.com/kylelemons/f9ed27f211dc818abc9b06cc3ff91540).